From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 6 Jun 2021 19:41:09 +0930
Subject: [PATCH] Configurable POI search delay


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 5c91c007691efeea1d4383efcd1ca06990d0660c..4e481d625e8ce957d58e66334cc3ef7e113198f7 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -832,4 +832,15 @@ public class PaperWorldConfig {
     private void fixInvulnerableEndCrystalExploit() {
         fixInvulnerableEndCrystalExploit = getBoolean("unsupported-settings.fix-invulnerable-end-crystal-exploit", fixInvulnerableEndCrystalExploit);
     }
+
+    public Map<EntityTypes<?>, Integer> poiSearchDelays = new HashMap<>();
+    private void poiSearchDelays() {
+        getInt("poi-search-delay.villager", 20);
+        EntityTypes.getEntityNameList().forEach(name -> {
+            final EntityTypes<?> type = EntityTypes.getByName(name.getKey()).orElseThrow(() -> new IllegalStateException("Unknown Entity Type: " + name.toString()));
+            final String path = ".poi-search-delay." + name.getKey();
+            final int value = config.getInt("world-settings." + worldName + path, config.getInt("world-settings.default" + path, -1)); // get without setting defaults
+            if (value != -1) poiSearchDelays.put(type, Math.max(1, value));
+        });
+    }
 }
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/BehaviorFindPosition.java b/src/main/java/net/minecraft/world/entity/ai/behavior/BehaviorFindPosition.java
index 1828ea10391e922d3ab3f5027a853af51000ca65..5bc1a911f9a501c22764e0278fcb0c9680990127 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/BehaviorFindPosition.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/BehaviorFindPosition.java
@@ -27,7 +27,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
     private final MemoryModuleType<GlobalPos> c;
     private final boolean d;
     private final Optional<Byte> e;
-    private long f;
+    private long f; private void setNextSearchTime(long time) {this.f = time;} // Paper - OBFHELPER
     private final Long2ObjectMap<BehaviorFindPosition.a> g;
 
     public BehaviorFindPosition(VillagePlaceType villageplacetype, MemoryModuleType<GlobalPos> memorymoduletype, MemoryModuleType<GlobalPos> memorymoduletype1, boolean flag, Optional<Byte> optional) {
@@ -58,7 +58,10 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
         if (this.d && entitycreature.isBaby()) {
             return false;
         } else if (this.f == 0L) {
-            this.f = entitycreature.world.getTime() + (long) worldserver.random.nextInt(20);
+            // Paper start
+            setNextSearchTime(entitycreature.world.getTime() + worldserver.random.nextInt(worldserver.paperConfig.poiSearchDelays.getOrDefault(entitycreature.getEntityType(), 20)));
+            //this.f = entitycreature.world.getTime() + (long) worldserver.random.nextInt(20);
+            // Paper end
             return false;
         } else {
             return worldserver.getTime() >= this.f;
@@ -66,7 +69,11 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
     }
 
     protected void a(WorldServer worldserver, EntityCreature entitycreature, long i) {
-        this.f = i + 20L + (long) worldserver.getRandom().nextInt(20);
+        // Paper start
+        int poiSearchDelay = worldserver.paperConfig.poiSearchDelays.getOrDefault(entitycreature.getEntityType(), 20);
+        setNextSearchTime(i + poiSearchDelay + worldserver.getRandom().nextInt(poiSearchDelay));
+        //this.f = i + 20L + (long) worldserver.getRandom().nextInt(20);
+        // Paper end
         VillagePlace villageplace = worldserver.y();
 
         this.g.long2ObjectEntrySet().removeIf((entry) -> {
